pallets/omnipool/src/traits.rs

```diff
@@ -143,17 +143,28 @@ where
 			return Ok(());
 		}
 
-		let external_price = ExternalOracle::get_price(asset_a, asset_b).map_err(|_| ())?;
-		let external_price = FixedU128::checked_from_rational(external_price.n, external_price.d).ok_or(())?;
 		let max_allowed = FixedU128::from(MaxAllowed::get());
-		let max_allowed_difference = external_price.saturating_mul(max_allowed);
+
+		let oracle_price = ExternalOracle::get_price(asset_a, asset_b).map_err(|_| ())?;
+		let external_price = FixedU128::checked_from_rational(oracle_price.n, oracle_price.d).ok_or(())?;
 		let current_spot_price = FixedU128::checked_from_rational(current_price.n, current_price.d).ok_or(())?;
 
+		let max_allowed_difference = max_allowed
+			.checked_mul(
+				&current_spot_price
+					.checked_add(&external_price)
+					.ok_or(())?
+					.checked_div(&FixedU128::from(2))
+					.ok_or(())?,
+			)
+			.ok_or(())?;
+
 		let diff = if current_spot_price >= external_price {
 			current_spot_price.saturating_sub(external_price)
 		} else {
 			external_price.saturating_sub(current_spot_price)
 		};
+
 		ensure!(diff <= max_allowed_difference, ());
```

Comment:
btw. you could pull out the `checked_div(2)` to here (multiplying it with `diff` instead)

Reply:
yes, good idea. so we avoid divisions where possible.